File: geonotebook/jsonrpc.py
import uuid


class JSONRPCError(Exception):
    code = 0
    message = None
    data = None
    status = 500

    def __init__(self, message=None):
        if message is not None:
            self.message = message

    def tojson(self):
        return {
            'name': self.__class__.__name__,
            'code': self.code,
            'message': "%s: %s" % (self.__class__.__name__, self.message),
            'data': self.data
        }


class ParseError(JSONRPCError):
    code = -32700
    message = "Parse Error."


class InvalidRequest(JSONRPCError):
    code = -32600
    message = "Invalid Request."


class MethodNotFound(JSONRPCError):
    code = -32601
    message = "Method not found."


class InvalidParams(JSONRPCError):
    code = -32602
    message = "Invalid params."


class InternalError(JSONRPCError):
    code = -32603
    message = "Internal Error."


class ServerError(JSONRPCError):
    code = -32000
    message = "Server Error."


def is_response(msg):
    return 'result' in msg and 'error' in msg and 'id' in msg


def is_request(msg):
    return 'method' in msg and 'params' in msg and 'id' in msg


def json_rpc_result(result, error, msg_id):
    return {
        "result": result,
        "error": error,
        "id": msg_id
    }


def json_rpc_request(method, params=None, jsonrpc="2.0"):
    return {
        "method": method,
        "params": params,
        "jsonrpc": jsonrpc,
        "id": str(uuid.uuid4())}


def json_rpc_notify(method, params=None, jsonrpc="2.0"):
    return {
        "method": method,
        "params": params,
        "jsonrpc": jsonrpc}

File: js/jsonrpc.js
import _ from 'underscore';

function JSONRPCError (code) {
  return function (message, data) {
    console.log('JSONRPCError (' + code + '): ' + message);

    if (data !== undefined) {
      return {
        code: code,
        message: message,
        data: data
      };
    } else {
      return {
        code: code,
        message: message
      };
    }
  };
}

function request (method, params, opts) {
  opts = _.defaults((typeof opts === 'undefined') ? {} : opts,
                          {jsonrpc: '2.0'});

  return _.extend({
    method: method,
    params: params}, opts);
}

function response (result, error, id) {
        // TODO validate taht result/error are mutually exclusive
  return {result: result,
    error: error,
    id: id };
}

function is_request (msg) {
  return 'method' in msg && 'params' in msg && 'id' in msg;
}

function is_response (msg) {
  return 'result' in msg && 'error' in msg && 'id' in msg;
}

const ParseError = JSONRPCError(-32700);
const InvalidRequest = JSONRPCError(-32600);
const MethodNotFound = JSONRPCError(-32601);
const InvalidParams = JSONRPCError(-32602);
const InternalError = JSONRPCError(-32603);
const ServerError = JSONRPCError(-32000);

export {
  JSONRPCError,
  request,
  response,
  is_request,
  is_response,
  ParseError,
  InvalidRequest,
  MethodNotFound,
  InvalidParams,
  InternalError,
  ServerError
};
